crux-lucene/src/crux/lucene.clj

```diff
@@ -0,0 +1,187 @@
+(ns crux.lucene
+  (:require [clojure.spec.alpha :as s]
+            [crux.codec :as cc]
+            [crux.db :as db]
+            [crux.io :as cio]
+            [crux.memory :as mem]
+            [crux.query :as q]
+            [crux.system :as sys]
+            [clojure.tools.logging :as log])
+  (:import crux.codec.EntityTx
+           java.io.Closeable
+           java.nio.file.Path
+           org.apache.lucene.analysis.Analyzer
+           org.apache.lucene.analysis.standard.StandardAnalyzer
+           [org.apache.lucene.document Document Field Field$Store StringField TextField]
+           [org.apache.lucene.index DirectoryReader IndexWriter IndexWriterConfig Term]
+           org.apache.lucene.queries.function.FunctionScoreQuery
+           org.apache.lucene.queryparser.classic.QueryParser
+           [org.apache.lucene.search BooleanClause$Occur BooleanQuery$Builder DoubleValuesSource IndexSearcher Query ScoreDoc TermQuery]
+           [org.apache.lucene.store Directory FSDirectory]))
+
+(def ^:dynamic *node*)
+
+(defrecord LuceneNode [directory analyzer]
+  java.io.Closeable
+  (close [this]
+    (doseq [^Closeable c [directory]]
+      (.close c))))
+
+(defn- id->stored-bytes [eid]
+  (mem/->on-heap (cc/->value-buffer eid)))
+
+(defn- ^String eid->str [eid]
+  (mem/buffer->hex (cc/->id-buffer eid)))
+
+(defn- crux-doc->triples [crux-doc]
+  (->> (dissoc crux-doc :crux.db/id)
+       (mapcat (fn [[k v]]
+                 (for [v (if (coll? v) v [v])
+                       :when (string? v)]
+                   [(name k) v])))))
+
+(defrecord DocumentId [a v])
+
+(defn- ^Document triple->doc [[k ^String v]]
+  (doto (Document.)
+    ;; To search for triples by eid-a-v for deduping
+    (.add (StringField. "id", (eid->str (DocumentId. k v)), Field$Store/NO))
+    ;; The actual term, which will be tokenized
+    (.add (TextField. (name k), v, Field$Store/YES))
+    ;; Uses for wildcard searches
+    (.add (TextField. "_val", v, Field$Store/YES))
+    ;; The Attr (storage only, for temporal resolution)
+    (.add (StringField. "_attr", (name k), Field$Store/YES))))
+
+(defn- ^Term triple->term [[k ^String v]]
+  (Term. "id" (eid->str (DocumentId. k v))))
+
+(defn doc-count []
+  (let [{:keys [^Directory directory]} *node*
+        directory-reader (DirectoryReader/open directory)]
+    (.numDocs directory-reader)))
+
+(defn write-docs! [^IndexWriter index-writer docs]
+  (doseq [d docs t (crux-doc->triples d)]
+    (.updateDocument index-writer (triple->term t) (triple->doc t))))
+
+(defn evict! [indexer, node, eids]
+  (let [{:keys [^Directory directory ^Analyzer analyzer]} node
+        attrs-id->attr (->> (db/read-index-meta indexer :crux/attribute-stats)
+                            keys
+                            (map #(vector (mem/buffer->hex (cc/->id-buffer %)) %))
```

Comment:
`(map (juxt eid->str identity))` ?
maybe `eid->str` should then be renamed to `->hex` or similar

Reply:
left as is, because it's not obvious hex is what's happening now